src/algorithms/LaurentPoly.jl

```diff
@@ -0,0 +1,212 @@
+###############################################################################
+#
+#   LaurentPoly.jl : Generic algorithms for abstract Laurent polynomials
+#
+###############################################################################
+
+# required methods without default implementation:
+# coeff, setcoeff!, map_coeffs, gen
+
+###############################################################################
+#
+#   Data type and parent object methods
+#
+###############################################################################
+
+base_ring(p::LaurentPolyElem) = base_ring(parent(p))
+
+isdomain_type(::Type{<:LaurentPolyElem{T}}) where {T} = isdomain_type(T)
+
+isexact_type(::Type{<:LaurentPolyElem{T}}) where {T} = isexact_type(T)
+
+function check_parent(a::LaurentPolyElem, b::LaurentPolyElem, throw::Bool = true)
+   c = parent(a) == parent(b)
+   c || !throw || error("incompatible Laurent polynomial rings")
+   return c
+end
+
+
+###############################################################################
+#
+#   Basic manipulation
+#
+###############################################################################
+
+const hashp_seed = UInt === UInt64 ? 0xcdaf0e0b5ade239b : 0x5ade239b
+
+function Base.hash(p::LaurentPolyElem, h::UInt)
+   for i in monomials_degrees(p)
+      c = coeff(p, i)
+      if !iszero(c)
+         h = hash(i, h)
+         h = hash(c, h)
+      end
+   end
+   hash(hashp_seed, h)
+end
+
+# required implementation
+"""
+    monomials_degrees(p::LaurentPolyElem) -> AbstractVector{<:Integer}
+
+> Return a vector containing at least all the degrees of the non-null
+> monomials of `p`.
```

Comment:
I think what you call a "monomial" is called a "term" elsewhere in AbstractAlgebra. But I might be wrong, I couldn't find a definition for what a "term" and what a "monomial" is, not even in the documentation for `isterm` and `ismonomial`.

These are unfortunately not universal. I know at least these conventions:
- `x^2` is a monomial and `2 x^2` is a term
- `x^2` is a term and `2 x^2` is a monomial
- `x^2` is a power product and `2 x^2` is a monomial or a term

Reply:
I think we use the first convention.

Reply:
Yes you are right, this function should be called `terms_degrees`, I will update it (and indeed, I'm very used to call this a monomial instead :) ).

Reply:
Convention 1 is also the one I usually use. We just should make sure to stay consistent with that. And ideally explicitly document the terminology somewhere -- e.g. in the `isterm` and `ismonomial` documentation. Of course that's not in the scope of this PR, just wanted to mention it.